# Cohort Building

**Example use case:** 

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" alt="alt_text" align="left"
	width="150" height="150" />
Julia is an oncologist that specializes in female reproductive health. As part of her research, she is interested in  using existing data on uterine cancers. If possible, she would like to see multiple datatypes (gross imaging, genomic data, proteomic data, histology) that come from the same patient, so she can look for shared phenotypes to test for their potential as early diagnostics. Julia heard that the Cancer Data Aggregator has made it easy to search across multiple datasets created by NCI, and so has decided to start her search there.



## Getting Started

The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:

---
- **<a href="../../QuickStart/usage/#q">Q.()</a>** builds a query that can be used by `run()` or `count()`
- **<a href="../../QuickStart/usage/#qrun">Q.run()</a>** returns data for the specified search 
- **<a href="../../QuickStart/usage/#qcount">Q.count()</a>** returns summary information (counts) data that fit the specified search
- **<a href="../../QuickStart/usage/#columns">columns()</a>** returns entity field names
- **<a href="../../QuickStart/usage/#unique_terms">unique_terms()</a>** returns entity field contents

---

Before Julia does any work, she needs to import these functions cdapython.
She'll also need to import [pandas](https://pandas.pydata.org/) to get nice dataframes.
Finally, she tells cdapython to report it's version so she can be sure she's using the one she means to:

In [1]:
from cdapython import Q, columns, unique_terms, query
import cdapython
import pandas as pd 
print(Q.get_version())

2022.6.28


In [2]:
Q.set_default_project_dataset("broad-dsde-dev.cda_dev")
Q.set_host_url("https://cancerdata.dsde-dev.broadinstitute.org/")
Q.get_host_url()
Q.get_default_project_dataset()

'broad-dsde-dev.cda_dev'

<div class="cdanote" style="background-color:#b3e5d5;color:black;padding:20px;">
    
    
CDA data comes from three sources:
<ul>
<li><b>The <a href="https://proteomic.datacommons.cancer.gov/pdc/"> Proteomic Data Commons</a> (PDC)</b></li>
<li><b>The <a href="https://gdc.cancer.gov/">Genomic Data Commons</a> (GDC)</b></li>
<li><b>The <a href="https://datacommons.cancer.gov/repository/imaging-data-commons">Imaging Data Commons</a> (IDC)</b></li>
</ul> 
    
The CDA makes this data searchable in four main endpoints:

<ul>
<li><b>subject:</b> A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</li>
<li><b>researchsubject:</b> A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</li>
<li><b>specimen:</b> Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</li>
<li><b>file:</b> A unit of data about subjects, researchsubjects, specimens, or their associated information</li>
</ul>
and two endpoints that offer deeper information about data in the researchsubject endpoint:
<ul>
<li><b>diagnosis:</b> A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</li>
<li><b>treatment:</b> Represent medication administration or other treatment types.</li>
</ul>
Any metadata field can be searched from any endpoint, the only difference between search types is what type of data is returned by default. This means that you can think of the CDA as a really, really enormous spreadsheet full of data. To search this enormous spreadsheet, you'd want select columns, and then filter rows.
</div>


## Finding Search Terms

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
   
   Accordingly, to see what search fields are available, Julia starts by using the command `columns`:

In [3]:
columns().to_list()

['id',
 'identifier.system',
 'identifier.value',
 'species',
 'sex',
 'race',
 'ethnicity',
 'days_to_birth',
 'subject_associated_project',
 'vital_status',
 'days_to_death',
 'cause_of_death',
 'ResearchSubject.id',
 'ResearchSubject.identifier.system',
 'ResearchSubject.identifier.value',
 'ResearchSubject.member_of_research_project',
 'ResearchSubject.primary_diagnosis_condition',
 'ResearchSubject.primary_diagnosis_site',
 'ResearchSubject.Diagnosis.id',
 'ResearchSubject.Diagnosis.identifier.system',
 'ResearchSubject.Diagnosis.identifier.value',
 'ResearchSubject.Diagnosis.primary_diagnosis',
 'ResearchSubject.Diagnosis.age_at_diagnosis',
 'ResearchSubject.Diagnosis.morphology',
 'ResearchSubject.Diagnosis.stage',
 'ResearchSubject.Diagnosis.grade',
 'ResearchSubject.Diagnosis.method_of_diagnosis',
 'ResearchSubject.Diagnosis.Treatment.id',
 'ResearchSubject.Diagnosis.Treatment.identifier.system',
 'ResearchSubject.Diagnosis.Treatment.identifier.value',
 'ResearchSubject.Diagno

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
   
There are a lot of columns in the CDA data, but Julia is most interested in diagnosis data, so she filters the list to only those:

In [4]:
columns().to_list(filters="diagnosis")

['ResearchSubject.primary_diagnosis_condition',
 'ResearchSubject.primary_diagnosis_site',
 'ResearchSubject.Diagnosis.id',
 'ResearchSubject.Diagnosis.identifier.system',
 'ResearchSubject.Diagnosis.identifier.value',
 'ResearchSubject.Diagnosis.primary_diagnosis',
 'ResearchSubject.Diagnosis.age_at_diagnosis',
 'ResearchSubject.Diagnosis.morphology',
 'ResearchSubject.Diagnosis.stage',
 'ResearchSubject.Diagnosis.grade',
 'ResearchSubject.Diagnosis.method_of_diagnosis',
 'ResearchSubject.Diagnosis.Treatment.id',
 'ResearchSubject.Diagnosis.Treatment.identifier.system',
 'ResearchSubject.Diagnosis.Treatment.identifier.value',
 'ResearchSubject.Diagnosis.Treatment.treatment_type',
 'ResearchSubject.Diagnosis.Treatment.treatment_outcome',
 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_start',
 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_end',
 'ResearchSubject.Diagnosis.Treatment.therapeutic_agent',
 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site',
 'Re

<div class="cdanote" style="background-color:#b3e5d5;color:black;padding:20px;">

To search the CDA, a user also needs to know what search terms are available. Each column will contain a huge amount of data, so retreiving all of the rows would be overwhelming. Instead, the CDA has a `unique_terms()` function that will return all of the unique values that populate the requested column. Like `columns`, `unique_terms` defaults to giving us an overview of the results, and can be filtered.
    
</div>

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
Since Julia is interested specificially in uterine cancers, she uses the `unique_terms` function to see what data is available for 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site' and 'ResearchSubject.primary_diagnosis_site' to see if 'uterine' appears:

In [5]:
unique_terms("ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site").to_list()

['Brain',
 'Cervix',
 'Head - Face Or Neck, Nos',
 'Lymph Node(s) Paraaortic',
 'Other',
 'Pelvis',
 'Spine',
 'Unknown']

In [6]:
unique_terms("ResearchSubject.primary_diagnosis_site").to_list()

['Abdomen',
 'Abdomen, Mediastinum',
 'Abdomen, Pelvis',
 'Adrenal Glands',
 'Adrenal gland',
 'Anus and anal canal',
 'Base of tongue',
 'Bile Duct',
 'Bladder',
 'Bones, joints and articular cartilage of limbs',
 'Bones, joints and articular cartilage of other and unspecified sites',
 'Brain',
 'Breast',
 'Bronchus and lung',
 'Cervix',
 'Cervix uteri',
 'Chest',
 'Chest-Abdomen-Pelvis, Leg, TSpine',
 'Colon',
 'Connective, subcutaneous and other soft tissues',
 'Corpus uteri',
 'Ear',
 'Esophagus',
 'Extremities',
 'Eye and adnexa',
 'Floor of mouth',
 'Gallbladder',
 'Gum',
 'Head',
 'Head and Neck',
 'Head-Neck',
 'Heart, mediastinum, and pleura',
 'Hematopoietic and reticuloendothelial systems',
 'Hypopharynx',
 'Intraocular',
 'Kidney',
 'Larynx',
 'Lip',
 'Liver',
 'Liver and intrahepatic bile ducts',
 'Lung',
 'Lung Phantom',
 'Lymph nodes',
 'Marrow, Blood',
 'Meninges',
 'Mesothelium',
 'Nasal cavity and middle ear',
 'Nasopharynx',
 'Not Reported',
 'Oropharynx',
 'Other an

<div class="cdanote" style="background-color:#b3e5d5;color:black;padding:20px;">
    
CDA makes multiple datasets searchable from a common interface, but does not harmonize the data. This means that researchers should review all the terms in a column, and not just choose the first one that fits, as there may be other similar terms available as well.
    
</div>

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
Julia sees that "treatment_anatomic_site" does not have 'Uterine', but does have 'Cervix'. She also notes that both 'Uterus' and 'Uterus, NOS' are listed in the "primary_diagnosis_site" results. As she was initially looking for "uterine", Julia decides to expand her search a bit to account for variable naming schemes. So, she runs a fuzzy match filter on the "ResearchSubject.primary_diagnosis_site" for 'uter' as that should cover all variants:

In [7]:
unique_terms("ResearchSubject.primary_diagnosis_site").to_list(filters="uter")

['Cervix uteri', 'Corpus uteri', 'Uterus', 'Uterus, NOS']

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
Just to be sure, Julia also searches for any other instances of "cervix":

In [8]:
unique_terms("ResearchSubject.primary_diagnosis_site").to_list(filters="cerv")

['Cervix', 'Cervix uteri']

## Building a Query

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
With all her likely terms found, Julia begins to create a search that will get data for all of her terms. She does this by writing a series of `Q` statements that define what rows should be returned from each column. For the "treatment_anatomic_site", only one term is of interest, so she uses the `=` operator to get only exact matches:

In [9]:
Tsite = Q('ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site = "Cervix"')

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
However, for "primary_diagnosis_site", Julia has several terms she wants to search with. Luckily, `Q` also can run fuzzy searches. It can also search more than one term at a time, so Julia writes one big `Q` statement to grab everything that is either 'uter' or 'cerv':

In [10]:
Dsite = Q('ResearchSubject.primary_diagnosis_site = "%uter%" OR ResearchSubject.primary_diagnosis_site = "%cerv%"')

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
Finally, Julia adds her two queries together into one large one:

In [11]:
ALLDATA = Tsite.OR(Dsite)

## Looking at Summary Data

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
Now that Julia has a query, she can use it to look for data in any of the CDA endpoints. She starts by getting an overall summary of what data is available using `count`:

In [12]:
ALLDATA.count.run()
#specimen_count : 40793
#treatment_count : 3049
#diagnosis_count : 3685
#researchsubject_count : 4869
#subject_count : 3742

Getting results from database

Total execution time: 3741 ms


specimen_count : 40793

treatment_count : 3049

diagnosis_count : 3685

researchsubject_count : 4869

subject_count : 3742

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
It seems there's a lot of data that might work for Julias study! Since she is interested in the beginings of cancer, she decides to start by looking at  the researchsubject information, since that is where most of the diagnosis information is. She again gets a summary using `count`:

In [13]:
ALLDATA.researchsubject.count.run()
# system	count
# GDC	3591
# PDC	104
# IDC	1174 
# primary_diagnosis_condition	count
# Adenomas and Adenocarcinomas	1672
# Uterine Corpus Endometrial Carcinoma	104
# Myomatous Neoplasms	188
# Cystic, Mucinous and Serous Neoplasms	487
# Squamous Cell Neoplasms	609
# Not Reported	12
# Complex Mixed and Stromal Neoplasms	320
# None	1175
# Epithelial Neoplasms, NOS	230
# Soft Tissue Tumors and Sarcomas, NOS	14
# Complex Epithelial Neoplasms	27
# Trophoblastic neoplasms	13
# Neoplasms, NOS	12
# Mesonephromas	5
# Neuroepitheliomatous Neoplasms	1 
# primary_diagnosis_site	count
# Cervix uteri	915
# Corpus uteri	780
# Uterus, NOS	2000
# Uterus	867
# Cervix	307

Getting results from database

Total execution time: 3562 ms


total : 4869

files : 324646

system 
 count 
 
 
 
 
 GDC 
 3591 
 
 
 PDC 
 104 
 
 
 IDC 
 1174 
 
 
 

 
 
 
 primary_diagnosis_condition 
 count 
 
 
 
 
 Adenomas and Adenocarcinomas 
 1672 
 
 
 Uterine Corpus Endometrial Carcinoma 
 104 
 
 
 Myomatous Neoplasms 
 188 
 
 
 Cystic, Mucinous and Serous Neoplasms 
 487 
 
 
 Squamous Cell Neoplasms 
 609 
 
 
 Not Reported 
 12 
 
 
 Complex Mixed and Stromal Neoplasms 
 320 
 
 
 None 
 1175 
 
 
 Epithelial Neoplasms, NOS 
 230 
 
 
 Soft Tissue Tumors and Sarcomas, NOS 
 14 
 
 
 Complex Epithelial Neoplasms 
 27 
 
 
 Trophoblastic neoplasms 
 13 
 
 
 Neoplasms, NOS 
 12 
 
 
 Mesonephromas 
 5 
 
 
 Neuroepitheliomatous Neoplasms 
 1 
 
 
 

 
 
 
 primary_diagnosis_site 
 count 
 
 
 
 
 Cervix uteri 
 915 
 
 
 Corpus uteri 
 780 
 
 
 Uterus, NOS 
 2000 
 
 
 Uterus 
 867 
 
 
 Cervix 
 307

## Refining Queries

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
Browsing the primary_diagnosis_condition data, Julia notices that there are a large number of research subjects that are Adenomas and Adenocarcinomas. Since Julia wants to look for common phenotypes in early cancers, she decides it might be easier to exclude the endocrine related data, as they might have different mechanisms. So she adds a new filter to her query:

In [14]:
Noadeno = Q('ResearchSubject.primary_diagnosis_condition != "Adenomas and Adenocarcinomas"')

NoAdenoData = ALLDATA.AND(Noadeno)

NoAdenoData.researchsubject.count.run()
#   total : 3197    
#    files : 298263   
# system	count
# PDC	104
# GDC	1919
# IDC	1174 
# primary_diagnosis_condition	count
# Myomatous Neoplasms	188
# Squamous Cell Neoplasms	609
# Uterine Corpus Endometrial Carcinoma	104
# Cystic, Mucinous and Serous Neoplasms	487
# Complex Mixed and Stromal Neoplasms	320
# None	1175
# Not Reported	12
# Epithelial Neoplasms, NOS	230
# Complex Epithelial Neoplasms	27
# Soft Tissue Tumors and Sarcomas, NOS	14
# Neuroepitheliomatous Neoplasms	1
# Trophoblastic neoplasms	13
# Neoplasms, NOS	12
# Mesonephromas	5 
# primary_diagnosis_site	count
# Cervix uteri	688
# Corpus uteri	373
# Uterus, NOS	962
# Uterus	867
# Cervix	307

Getting results from database

Total execution time: 3560 ms


total : 3197

files : 298263

system 
 count 
 
 
 
 
 PDC 
 104 
 
 
 GDC 
 1919 
 
 
 IDC 
 1174 
 
 
 

 
 
 
 primary_diagnosis_condition 
 count 
 
 
 
 
 Myomatous Neoplasms 
 188 
 
 
 Squamous Cell Neoplasms 
 609 
 
 
 Uterine Corpus Endometrial Carcinoma 
 104 
 
 
 Cystic, Mucinous and Serous Neoplasms 
 487 
 
 
 Complex Mixed and Stromal Neoplasms 
 320 
 
 
 None 
 1175 
 
 
 Not Reported 
 12 
 
 
 Epithelial Neoplasms, NOS 
 230 
 
 
 Complex Epithelial Neoplasms 
 27 
 
 
 Soft Tissue Tumors and Sarcomas, NOS 
 14 
 
 
 Neuroepitheliomatous Neoplasms 
 1 
 
 
 Trophoblastic neoplasms 
 13 
 
 
 Neoplasms, NOS 
 12 
 
 
 Mesonephromas 
 5 
 
 
 

 
 
 
 primary_diagnosis_site 
 count 
 
 
 
 
 Cervix uteri 
 688 
 
 
 Corpus uteri 
 373 
 
 
 Uterus, NOS 
 962 
 
 
 Uterus 
 867 
 
 
 Cervix 
 307

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
She then previews the actual metadata for researchsubject, subject, and file, to make sure that they have all the information she will need for her work. Since she's mostly interested in looking at the kinds of data available from each endpoint:

In [15]:
NoAdenoData.researchsubject.run().to_dataframe() # view the dataframe

Getting results from database

Total execution time: 3622 ms


id  \
0   1f20b0c8-11c1-4a3c-84fc-d485aac49bc8   
1   29d93fb1-0b3d-4d13-8799-2dcf3e14be04   
2   3bccf0e3-d467-4477-adfd-b04d71f3eb86   
3   3c25bf98-1a11-41dc-8d65-182b0e8c5978   
4   4ed1616e-8c62-4104-bcbe-26d7059f04d0   
..                                   ...   
95  3789ae3f-286b-42f0-a805-320ad69ea1df   
96  3ac62513-b5f4-4063-aabd-dad08a1f56fb   
97  3ba89750-e310-4771-88e7-b5bf289354e0   
98  5040ec9a-5050-4295-b792-38863eded12b   
99  5be58e2e-118a-11e9-afb9-0a9c39d33490   

                                           identifier  \
0   [{'system': 'GDC', 'value': '1f20b0c8-11c1-4a3...   
1   [{'system': 'GDC', 'value': '29d93fb1-0b3d-4d1...   
2   [{'system': 'GDC', 'value': '3bccf0e3-d467-447...   
3   [{'system': 'GDC', 'value': '3c25bf98-1a11-41d...   
4   [{'system': 'GDC', 'value': '4ed1616e-8c62-410...   
..                                                ...   
95  [{'system': 'GDC', 'value': '3789ae3f-286b-42f...   
96  [{'system': 'GDC', 'value': '3ac62513-b5f4-406...   
97  [{'system': 'GDC', 'value': '3ba89750-e310-477...   
98  [{'system': 'GDC', 'value': '5040ec9a-5050-429...   
99  [{'system': 'PDC', 'value': '5be58e2e-118a-11e...   

   member_of_research_project            primary_diagnosis_condition  \
0                       FM-AD    Complex Mixed and Stromal Neoplasms   
1                  GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   
2                  GENIE-DFCI                    Myomatous Neoplasms   
3                   GENIE-MSK                    Myomatous Neoplasms   
4                   TCGA-CESC                Squamous Cell Neoplasms   
..                        ...                                    ...   
95                  TCGA-UCEC  Cystic, Mucinous and Serous Neoplasms   
96                  TCGA-CESC                Squamous Cell Neoplasms   
97                      FM-AD                Squamous Cell Neoplasms   
98                      FM-AD                Squamous Cell Neoplasms   
99           CPTAC3-Discovery   Uterine Corpus Endometrial Carcinoma   

   primary_diagnosis_site           subject_id  
0             Uterus, NOS              AD16769  
1             Uterus, NOS    GENIE-DFCI-006032  
2             Uterus, NOS    GENIE-DFCI-009614  
3             Uterus, NOS  GENIE-MSK-P-0014679  
4            Cervix uteri         TCGA-C5-A7X3  
..                    ...                  ...  
95           Corpus uteri         TCGA-B5-A3S1  
96           Cervix uteri         TCGA-EA-A3Y4  
97           Cervix uteri              AD16507  
98           Cervix uteri               AD3542  
99            Uterus, NOS            C3L-00963  

[100 rows x 6 columns]

---

<div class="cdadefine" style="background-color:#add9e5;color:black;padding:20px;">

<h3>ResearchSubject Field Definitions</h3>

<i>A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</i>
    
<ul>
<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. For CDA, this is case_id.</li>
<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>
<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>
<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>
<li><b>member_of_research_project:</b> A reference to the Study(s) of which this ResearchSubject is a member.</li>
<li><b>primary_diagnosis_condition:</b> The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) <a href="https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology">International Classification of Diseases for Oncology</a> (ICD-O). This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.</li>
<li><b>primary_diagnosis_site:</b> The text term used to describe the primary site of disease, as categorized by the World Health Organization's (WHO) <a href="https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology">International Classification of Diseases for Oncology</a> (ICD-O). This categorization groups cases into general categories. This attribute represents the primary site of disease that qualified the subject for inclusion on the ResearchProject.</li>
<li><b>subject_id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. Can be joined to the `id` field from subject results</li>
</ul>  

</div>
    
---

In [16]:
NoAdenoData.subject.run().to_dataframe() # view the dataframe

Getting results from database

Total execution time: 3596 ms


id                                         identifier  \
0                AD95               [{'system': 'GDC', 'value': 'AD95'}]   
1           C3L-02894          [{'system': 'IDC', 'value': 'C3L-02894'}]   
2           C3N-03005          [{'system': 'IDC', 'value': 'C3N-03005'}]   
3   GENIE-DFCI-002257  [{'system': 'GDC', 'value': 'GENIE-DFCI-002257'}]   
4   GENIE-DFCI-003511  [{'system': 'GDC', 'value': 'GENIE-DFCI-003511'}]   
..                ...                                                ...   
95       TCGA-EY-A3QX  [{'system': 'GDC', 'value': 'TCGA-EY-A3QX'}, {...   
96       TCGA-N5-A4RU  [{'system': 'GDC', 'value': 'TCGA-N5-A4RU'}, {...   
97       TCGA-NG-A4VU  [{'system': 'GDC', 'value': 'TCGA-NG-A4VU'}, {...   
98       TCGA-PG-A7D5  [{'system': 'GDC', 'value': 'TCGA-PG-A7D5'}, {...   
99       TCGA-VS-A959  [{'system': 'GDC', 'value': 'TCGA-VS-A959'}, {...   

         species     sex                       race               ethnicity  \
0   homo sapiens  female               not reported            not reported   
1   Homo sapiens    None                       None                    None   
2   Homo sapiens    None                       None                    None   
3   homo sapiens  female                      white  not hispanic or latino   
4   homo sapiens  female                      white  not hispanic or latino   
..           ...     ...                        ...                     ...   
95  homo sapiens  female                      white  not hispanic or latino   
96  homo sapiens  female                      white  not hispanic or latino   
97  homo sapiens  female                      white  not hispanic or latino   
98  homo sapiens  female  black or african american  not hispanic or latino   
99  homo sapiens  female               not reported            not reported   

    days_to_birth subject_associated_project  vital_status  days_to_death  \
0             NaN                    [FM-AD]  Not Reported            NaN   
1             NaN               [cptac_ucec]          None            NaN   
2             NaN               [cptac_ucec]          None            NaN   
3        -22280.0               [GENIE-DFCI]  Not Reported            NaN   
4        -24837.0               [GENIE-DFCI]  Not Reported            NaN   
..            ...                        ...           ...            ...   
95       -23663.0     [TCGA-UCEC, tcga_ucec]          Dead          989.0   
96       -20025.0       [TCGA-UCS, tcga_ucs]          Dead         1526.0   
97       -23323.0       [TCGA-UCS, tcga_ucs]          Dead          442.0   
98       -23006.0     [TCGA-UCEC, tcga_ucec]         Alive            NaN   
99       -27897.0     [TCGA-CESC, tcga_cesc]         Alive            NaN   

   cause_of_death  
0            None  
1            None  
2            None  
3            None  
4            None  
..            ...  
95           None  
96           None  
97           None  
98           None  
99           None  

[100 rows x 11 columns]

---

<div class="cdadefine" style="background-color:#add9e5;color:black;padding:20px;">

<h3>Subject Field Definitions</h3>

<i>A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</i>

    
<ul>
<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.",STRING</li>
<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>
<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>
<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>
<li><b>species:</b> The taxonomic group (e.g. species) of the patient. For MVP, since taxonomy vocabulary is consistent between GDC and PDC, using text. Ultimately, this will be a term returned by the vocabulary service.</li>
<li><b>sex:</b> The biologic character or quality that distinguishes male and female from one another as expressed by analysis of the person's gonadal, morphologic (internal and external), chromosomal, and hormonal characteristics.</li>
<li><b>race:</b> An arbitrary classification of a taxonomic group that is a division of a species. It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>
<li><b>ethnicity:</b> An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>
<li><b>days_to_birth:</b> Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days.</li>
<li><b>subject_associated_project:</b> The list of Projects associated with the Subject.</li>
<li><b>vital_status:</b> Coded value indicating the state or condition of being living or deceased; also includes the case where the vital status is unknown.</li>
<li><b>days_to_death:</b> Number of days between the date used for index and the date from a person's date of death represented as a calculated number of days.</li>
<li><b>cause_of_death:</b> Coded value indicating the circumstance or condition that results in the death of the subject.</li>
</ul>  

</div>
    
---

In [17]:
NoAdenoData.file.run().to_dataframe() # view the dataframe

Getting results from database

Total execution time: 4225 ms


id  \
0   0b3ca53a-102f-4e81-845f-cad1afeabcb9   
1   308d0d13-ba05-46b4-a46d-ec02e9768f63   
2   a1e661f5-3158-4748-8e4c-c262f6981e05   
3   a6de4f29-79fb-4139-b320-b664d2d5387c   
4   b30cd89d-895b-4d7c-b2dc-9cb469c9b39f   
..                                   ...   
95  fed0013e-e5e2-4fce-b1f1-e7bc942b4a4a   
96  0b5cd4dc-5081-466a-9bcc-bc717b3254af   
97  0f6bff1c-a0e0-47f6-b4a9-6eff171b7e82   
98  1844c0ab-f430-45bc-9865-4078cf57c5c4   
99  1981e3bd-c534-4259-8124-cd37a51fb725   

                                           identifier  \
0   [{'system': 'GDC', 'value': '0b3ca53a-102f-4e8...   
1   [{'system': 'GDC', 'value': '308d0d13-ba05-46b...   
2   [{'system': 'GDC', 'value': 'a1e661f5-3158-474...   
3   [{'system': 'GDC', 'value': 'a6de4f29-79fb-413...   
4   [{'system': 'GDC', 'value': 'b30cd89d-895b-4d7...   
..                                                ...   
95  [{'system': 'GDC', 'value': 'fed0013e-e5e2-4fc...   
96  [{'system': 'GDC', 'value': '0b5cd4dc-5081-466...   
97  [{'system': 'GDC', 'value': '0f6bff1c-a0e0-47f...   
98  [{'system': 'GDC', 'value': '1844c0ab-f430-45b...   
99  [{'system': 'GDC', 'value': '1981e3bd-c534-425...   

                                                label  \
0   baff6124-e0ed-4ae8-abb2-614fc68e6728.wgs.BRASS...   
1   32ac189b-859b-4f44-8040-c68b5bfc499d.wgs.BRASS...   
2           TCGA-2W-A8YY-01A-21-A40H-20_RPPA_data.tsv   
3   16afeee8-917a-4682-abf6-3d599345fc28.wgs.BRASS...   
4   c2ad8dfa-b36b-4bf8-8db2-1a3beb9ec181.wgs.BRASS...   
..                                                ...   
95  a415048b-118c-4e8e-a44d-ca9ea524eedd.wgs.BRASS...   
96  2db011f6-0005-4fab-ad90-a0e02afc1a0b.wgs.BRASS...   
97  e05cb0b1-3d3b-47d0-b513-aeeba47b313d.wgs.BRASS...   
98  bcdf7a5f-2726-4b83-9a56-0ddf6e5d48c4.wgs.BRASS...   
99  d6381f8b-f7e0-49c6-9b1b-481079e84410.wgs.BRASS...   

                   data_category                          data_type  \
0   Somatic Structural Variation           Structural Rearrangement   
1   Somatic Structural Variation           Structural Rearrangement   
2             Proteome Profiling  Protein Expression Quantification   
3   Somatic Structural Variation           Structural Rearrangement   
4   Somatic Structural Variation           Structural Rearrangement   
..                           ...                                ...   
95  Somatic Structural Variation           Structural Rearrangement   
96  Somatic Structural Variation           Structural Rearrangement   
97  Somatic Structural Variation           Structural Rearrangement   
98  Somatic Structural Variation           Structural Rearrangement   
99  Somatic Structural Variation           Structural Rearrangement   

   file_format associated_project  \
0        BEDPE      CGCI-HTMCP-CC   
1          VCF      CGCI-HTMCP-CC   
2          TSV          TCGA-CESC   
3          VCF      CGCI-HTMCP-CC   
4          VCF      CGCI-HTMCP-CC   
..         ...                ...   
95         VCF      CGCI-HTMCP-CC   
96         VCF      CGCI-HTMCP-CC   
97       BEDPE      CGCI-HTMCP-CC   
98       BEDPE      CGCI-HTMCP-CC   
99         VCF      CGCI-HTMCP-CC   

                                              drs_uri  byte_size  \
0   drs://dg.4DFC:0b3ca53a-102f-4e81-845f-cad1afea...     186898   
1   drs://dg.4DFC:308d0d13-ba05-46b4-a46d-ec02e976...      64370   
2   drs://dg.4DFC:a1e661f5-3158-4748-8e4c-c262f698...      23833   
3   drs://dg.4DFC:a6de4f29-79fb-4139-b320-b664d2d5...      38046   
4   drs://dg.4DFC:b30cd89d-895b-4d7c-b2dc-9cb469c9...      48137   
..                                                ...        ...   
95  drs://dg.4DFC:fed0013e-e5e2-4fce-b1f1-e7bc942b...     109420   
96  drs://dg.4DFC:0b5cd4dc-5081-466a-9bcc-bc717b32...      67148   
97  drs://dg.4DFC:0f6bff1c-a0e0-47f6-b4a9-6eff171b...     106577   
98  drs://dg.4DFC:1844c0ab-f430-45bc-9865-4078cf57...      87697   
99  drs://dg.4DFC:1981e3bd-c534-4259-8124-cd37a51f...      60710   

                        


---

<div class="cdadefine" style="background-color:#add9e5;color:black;padding:20px;">

<h3>File Field Definitions</h3>

<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>

    
<ul>
  <li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>
<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>
<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>
<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>
  <li><b>label:</b> Short name or abbreviation for dataset. Maps to rdfs:label.</li>
  <li><b>data_catagory:</b> Broad categorization of the contents of the data file.</li>
  <li><b>data_type:</b> Specific content type of the data file.</li>
  <li><b>file_format:</b> Format of the data files.</li>
  <li><b>associated_project:</b> A reference to the Project(s) of which this ResearchSubject is a member. The associated_project may be embedded using the ref definition or may be a reference to the id for the Project - or a URI expressed as a string to an existing entity.</li>
  <li><b>drs_uri:</b> A string of characters used to identify a resource on the Data Repo Service(DRS). Can be used to retreive this specific file from a server.</li>
  <li><b>byte_size:</b> Size of the file in bytes. Maps to dcat:byteSize.</li>
  <li><b>checksum:</b> The md5 value for the file. A digit representing the sum of the correct digits in a piece of stored or transmitted digital data, against which later comparisons can be made to detect errors in the data.</li>
  <li><b>data_modality:</b> Data modality describes the biological nature of the information gathered as the result of an Activity, independent of the technology or methods used to produce the information. Always one of "Genomic", "Proteomic", or "Imaging".</li>
  <li><b>imaging_modality:</b> An imaging modality describes the imaging equipment and/or method used to acquire certain structural or functional information about the body. These include but are not limited to computed tomography (CT) and magnetic resonance imaging (MRI). Taken from the DICOM standard.</li>
  <li><b>dbgap_accession_number:</b> The dbgap accession number for the project.</li>
</ul>  

</div>
    
---


## Working with Results (pagination)

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
Finally, Julia wants to save these results to use for the future. Since the preview dataframes only show the first 100 results of each search, she uses the `paginator` function to get all the data from the subject and researchsubject endpoints into their own dataframes:

In [18]:
researchsubs = NoAdenoData.researchsubject.run()
rsdf = pd.DataFrame()
for i in researchsubs.paginator(to_df=True):
    rsdf = pd.concat([rsdf, i])

Getting results from database

Total execution time: 3592 ms


In [19]:
subs = NoAdenoData.subject.run()
subsdf = pd.DataFrame()
for i in subs.paginator(to_df=True):
    subsdf = pd.concat([subsdf, i])

Getting results from database

Total execution time: 3596 ms


In [20]:
rsdf # view the researchsubject dataframe

id  \
0   1f20b0c8-11c1-4a3c-84fc-d485aac49bc8   
1   29d93fb1-0b3d-4d13-8799-2dcf3e14be04   
2   3bccf0e3-d467-4477-adfd-b04d71f3eb86   
3   3c25bf98-1a11-41dc-8d65-182b0e8c5978   
4   4ed1616e-8c62-4104-bcbe-26d7059f04d0   
..                                   ...   
92  c2b54dd8-ef0a-464a-b31e-3a4c1e9cd20c   
93  d5ee39e8-a1d2-4244-9c38-07d7c5c1baba   
94  e5e66e4f-46af-4d88-a238-a8eb8e9dca09   
95  f8c840a8-2891-4c72-b9d8-9eb9c4351cf7   
96  fbd1ffac-d944-4356-b224-663696112a0b   

                                           identifier  \
0   [{'system': 'GDC', 'value': '1f20b0c8-11c1-4a3...   
1   [{'system': 'GDC', 'value': '29d93fb1-0b3d-4d1...   
2   [{'system': 'GDC', 'value': '3bccf0e3-d467-447...   
3   [{'system': 'GDC', 'value': '3c25bf98-1a11-41d...   
4   [{'system': 'GDC', 'value': '4ed1616e-8c62-410...   
..                                                ...   
92  [{'system': 'GDC', 'value': 'c2b54dd8-ef0a-464...   
93  [{'system': 'GDC', 'value': 'd5ee39e8-a1d2-424...   
94  [{'system': 'GDC', 'value': 'e5e66e4f-46af-4d8...   
95  [{'system': 'GDC', 'value': 'f8c840a8-2891-4c7...   
96  [{'system': 'GDC', 'value': 'fbd1ffac-d944-435...   

   member_of_research_project            primary_diagnosis_condition  \
0                       FM-AD    Complex Mixed and Stromal Neoplasms   
1                  GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   
2                  GENIE-DFCI                    Myomatous Neoplasms   
3                   GENIE-MSK                    Myomatous Neoplasms   
4                   TCGA-CESC                Squamous Cell Neoplasms   
..                        ...                                    ...   
92                  GENIE-MSK              Epithelial Neoplasms, NOS   
93                  TCGA-UCEC  Cystic, Mucinous and Serous Neoplasms   
94                 GENIE-DFCI              Epithelial Neoplasms, NOS   
95                  GENIE-UHN  Cystic, Mucinous and Serous Neoplasms   
96                  GENIE-MSK              Epithelial Neoplasms, NOS   

   primary_diagnosis_site           subject_id  
0             Uterus, NOS              AD16769  
1             Uterus, NOS    GENIE-DFCI-006032  
2             Uterus, NOS    GENIE-DFCI-009614  
3             Uterus, NOS  GENIE-MSK-P-0014679  
4            Cervix uteri         TCGA-C5-A7X3  
..                    ...                  ...  
92            Uterus, NOS  GENIE-MSK-P-0000080  
93           Corpus uteri         TCGA-B5-A5OE  
94           Corpus uteri    GENIE-DFCI-037165  
95            Uterus, NOS     GENIE-UHN-242802  
96           Corpus uteri  GENIE-MSK-P-0012305  

[3197 rows x 6 columns]

In [21]:
subsdf # view the subject dataframe

id                                         identifier  \
0                AD95               [{'system': 'GDC', 'value': 'AD95'}]   
1           C3L-02894          [{'system': 'IDC', 'value': 'C3L-02894'}]   
2           C3N-03005          [{'system': 'IDC', 'value': 'C3N-03005'}]   
3   GENIE-DFCI-002257  [{'system': 'GDC', 'value': 'GENIE-DFCI-002257'}]   
4   GENIE-DFCI-003511  [{'system': 'GDC', 'value': 'GENIE-DFCI-003511'}]   
..                ...                                                ...   
4        TCGA-D1-A1NZ  [{'system': 'GDC', 'value': 'TCGA-D1-A1NZ'}, {...   
5        TCGA-D1-A2G5  [{'system': 'GDC', 'value': 'TCGA-D1-A2G5'}, {...   
6        TCGA-IS-A3K7  [{'system': 'GDC', 'value': 'TCGA-IS-A3K7'}, {...   
7        TCGA-NF-A4WU  [{'system': 'GDC', 'value': 'TCGA-NF-A4WU'}, {...   
8        TCGA-X2-A95T  [{'system': 'GDC', 'value': 'TCGA-X2-A95T'}, {...   

         species     sex          race               ethnicity  days_to_birth  \
0   homo sapiens  female  not reported            not reported            NaN   
1   Homo sapiens    None          None                    None            NaN   
2   Homo sapiens    None          None                    None            NaN   
3   homo sapiens  female         white  not hispanic or latino       -22280.0   
4   homo sapiens  female         white  not hispanic or latino       -24837.0   
..           ...     ...           ...                     ...            ...   
4   homo sapiens  female         white  not hispanic or latino       -22027.0   
5   homo sapiens  female         white  not hispanic or latino       -18322.0   
6   homo sapiens  female         white            not reported       -23022.0   
7   homo sapiens  female         white  not hispanic or latino       -21901.0   
8   homo sapiens  female         white            not reported       -18079.0   

   subject_associated_project  vital_status  days_to_death cause_of_death  
0                     [FM-AD]  Not Reported            NaN           None  
1                [cptac_ucec]          None            NaN           None  
2                [cptac_ucec]          None            NaN           None  
3                [GENIE-DFCI]  Not Reported            NaN           None  
4                [GENIE-DFCI]  Not Reported            NaN           None  
..                        ...           ...            ...            ...  
4      [TCGA-UCEC, tcga_ucec]         Alive            NaN           None  
5      [TCGA-UCEC, tcga_ucec]         Alive            NaN           None  
6      [tcga_sarc, TCGA-SARC]         Alive            NaN           None  
7        [TCGA-UCS, tcga_ucs]         Alive            NaN           None  
8      [tcga_sarc, TCGA-SARC]         Alive            NaN           None  

[2609 rows x 11 columns]

## Merging Results across Endpoints

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
Then Julia uses the `id` fields in each result to merge them together into one big dataset:

In [22]:
allmetadata = pd.merge(rsdf,
                subsdf,
                left_on="subject_id",
                right_on='id')

allmetadata

id_x  \
0     1f20b0c8-11c1-4a3c-84fc-d485aac49bc8   
1     29d93fb1-0b3d-4d13-8799-2dcf3e14be04   
2     3bccf0e3-d467-4477-adfd-b04d71f3eb86   
3     3c25bf98-1a11-41dc-8d65-182b0e8c5978   
4     4ed1616e-8c62-4104-bcbe-26d7059f04d0   
...                                    ...   
3192  baf1b1f8-36ae-4370-98dd-23bda7ed147e   
3193  c2b54dd8-ef0a-464a-b31e-3a4c1e9cd20c   
3194  e5e66e4f-46af-4d88-a238-a8eb8e9dca09   
3195  f8c840a8-2891-4c72-b9d8-9eb9c4351cf7   
3196  fbd1ffac-d944-4356-b224-663696112a0b   

                                           identifier_x  \
0     [{'system': 'GDC', 'value': '1f20b0c8-11c1-4a3...   
1     [{'system': 'GDC', 'value': '29d93fb1-0b3d-4d1...   
2     [{'system': 'GDC', 'value': '3bccf0e3-d467-447...   
3     [{'system': 'GDC', 'value': '3c25bf98-1a11-41d...   
4     [{'system': 'GDC', 'value': '4ed1616e-8c62-410...   
...                                                 ...   
3192  [{'system': 'GDC', 'value': 'baf1b1f8-36ae-437...   
3193  [{'system': 'GDC', 'value': 'c2b54dd8-ef0a-464...   
3194  [{'system': 'GDC', 'value': 'e5e66e4f-46af-4d8...   
3195  [{'system': 'GDC', 'value': 'f8c840a8-2891-4c7...   
3196  [{'system': 'GDC', 'value': 'fbd1ffac-d944-435...   

     member_of_research_project            primary_diagnosis_condition  \
0                         FM-AD    Complex Mixed and Stromal Neoplasms   
1                    GENIE-DFCI  Cystic, Mucinous and Serous Neoplasms   
2                    GENIE-DFCI                    Myomatous Neoplasms   
3                     GENIE-MSK                    Myomatous Neoplasms   
4                     TCGA-CESC                Squamous Cell Neoplasms   
...                         ...                                    ...   
3192                  GENIE-MSK  Cystic, Mucinous and Serous Neoplasms   
3193                  GENIE-MSK              Epithelial Neoplasms, NOS   
3194                 GENIE-DFCI              Epithelial Neoplasms, NOS   
3195                  GENIE-UHN  Cystic, Mucinous and Serous Neoplasms   
3196                  GENIE-MSK              Epithelial Neoplasms, NOS   

     primary_diagnosis_site           subject_id                 id_y  \
0               Uterus, NOS              AD16769              AD16769   
1               Uterus, NOS    GENIE-DFCI-006032    GENIE-DFCI-006032   
2               Uterus, NOS    GENIE-DFCI-009614    GENIE-DFCI-009614   
3               Uterus, NOS  GENIE-MSK-P-0014679  GENIE-MSK-P-0014679   
4              Cervix uteri         TCGA-C5-A7X3         TCGA-C5-A7X3   
...                     ...                  ...                  ...   
3192            Uterus, NOS  GENIE-MSK-P-0005695  GENIE-MSK-P-0005695   
3193            Uterus, NOS  GENIE-MSK-P-0000080  GENIE-MSK-P-0000080   
3194           Corpus uteri    GENIE-DFCI-037165    GENIE-DFCI-037165   
3195            Uterus, NOS     GENIE-UHN-242802     GENIE-UHN-242802   
3196           Corpus uteri  GENIE-MSK-P-0012305  GENIE-MSK-P-0012305   

                                           identifier_y       species     sex  \
0               [{'system': 'GDC', 'value': 'AD16769'}]  homo sapiens  female   
1     [{'system': 'GDC', 'value': 'GENIE-DFCI-006032'}]  homo sapiens  female   
2     [{'system': 'GDC', 'value': 'GENIE-DFCI-009614'}]  homo sapiens  female   
3     [{'system': 'GDC', 'value': 'GENIE-MSK-P-00146...  homo sapiens  female   
4     [{'system': 'GDC', 'value': 'TCGA-C5-A7X3'}, {...  homo sapiens  female   
...                                                 ...           ...     ...   
3192  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00056...  homo sapiens  female   
3193  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00000...  homo sapiens  female   
3194  [{'system': 'GDC', 'value': 'GENIE-DFCI-037165'}]  homo sapiens  female   
3195   [{'system': 'GDC', 'value': 'GENIE-UHN-242802'}]  homo sapiens  female   
3196  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00123...  homo sapiens  female   

                           race      

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
And saves it out to a csv so she can browse it with Excel:

In [23]:
allmetadata.to_csv("allmetadata.csv")

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
   
Julia knows from her subject count summary that there are more than 200,000 files associated with her subjects, which is likely far more than she needs. To help her decide what files she wants, Julia uses endpoint chaining to get summary information about the files that are assigned to researchsubjects for her search criteria:


In [24]:
NoAdenoData.researchsubject.file.count.run()

Getting results from database

Total execution time: 3505 ms


total : 298263

system 
 count 
 
 
 
 
 GDC 
 31274 
 
 
 IDC 
 264429 
 
 
 PDC 
 2560 
 
 
 

 
 
 
 data_category 
 count 
 
 
 
 
 Imaging 
 264429 
 
 
 Processed Mass Spectra 
 640 
 
 
 Simple Nucleotide Variation 
 11745 
 
 
 Copy Number Variation 
 4079 
 
 
 Sequencing Reads 
 4142 
 
 
 Clinical 
 774 
 
 
 Somatic Structural Variation 
 370 
 
 
 Biospecimen 
 2866 
 
 
 Structural Variation 
 2192 
 
 
 Raw Mass Spectra 
 640 
 
 
 Peptide Spectral Matches 
 1280 
 
 
 Transcriptome Profiling 
 2820 
 
 
 DNA Methylation 
 1947 
 
 
 Proteome Profiling 
 339 
 
 
 

 
 
 
 file_format 
 count 
 
 
 
 
 DICOM 
 264429 
 
 
 BAM 
 4142 
 
 
 VCF 
 6652 
 
 
 SVS 
 1111 
 
 
 tsv 
 640 
 
 
 TSV 
 3980 
 
 
 BEDPE 
 1504 
 
 
 BCR SSF XML 
 517 
 
 
 MAF 
 5235 
 
 
 vendor-specific 
 640 
 
 
 TXT 
 4827 
 
 
 IDAT 
 1298 
 
 
 BCR XML 
 1217 
 
 
 BCR Auxiliary XML 
 474 
 
 
 mzIdentML 
 640 
 
 
 mzML 
 640 
 
 
 BCR PPS XML 
 193 
 
 
 BCR OMF XML 
 39 
 
 
 BCR Biotab 
 76 
 
 
 XLSX 
 7 
 
 
 PDF 
 1 
 
 
 CDC JSON 
 1 
 
 
 

 
 
 
 data_type 
 count 
 
 
 
 
 Annotated Somatic Mutation 
 6043 
 
 
 None 
 264429 
 
 
 Aggregated Somatic Mutation 
 732 
 
 
 Isoform Expression Quantification 
 700 
 
 
 Slide Image 
 1111 
 
 
 Copy Number Segment 
 1140 
 
 
 Proprietary 
 640 
 
 
 Masked Intensities 
 1298 
 
 
 Aligned Reads 
 4142 
 
 
 Gene Level Copy Number Scores 
 809 
 
 
 miRNA Expression Quantification 
 700 
 
 
 Biospecimen Supplement 
 1755 
 
 
 Transcript Fusion 
 2278 
 
 
 Protein Expression Quantification 
 339 
 
 
 Masked Copy Number Segment 
 998 
 
 
 Raw Simple Somatic Mutation 
 3397 
 
 
 Splice Junction Quantification 
 710 
 
 
 Open Standard 
 1280 
 
 
 Gene Expression Quantification 
 710 
 
 
 Structural Rearrangement 
 284 
 
 
 Masked Annotated Somatic Mutation 
 1024 
 
 
 Gene Level Copy Number 
 637 
 
 
 Methylation Beta Value 
 649 
 
 
 Masked Somatic Mutation 
 549 
 
 
 Clinical Supplement 
 773 
 
 
 Allele-specific Copy Number Segment 
 495 
 
 
 Text 
 640 
 
 
 Pathology Report 
 1

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
    
Julia decides that a good place to start would be with Slide Images. There's only 1111, so she should be able to quickly scan through them over the next few days and see if they will be useful. So she adds one more filter on her search:

In [25]:
JustSlides = Q('file.data_type = "Slide Image"')
NoadenoJustSlides = NoAdenoData.AND(JustSlides)
NoadenoJustSlides.researchsubject.file.count.run()

Getting results from database

Total execution time: 3618 ms


total : 1111

system 
 count 
 
 
 
 
 GDC 
 1111 
 
 
 

 
 
 
 data_category 
 count 
 
 
 
 
 Biospecimen 
 1111 
 
 
 

 
 
 
 file_format 
 count 
 
 
 
 
 SVS 
 1111 
 
 
 

 
 
 
 data_type 
 count 
 
 
 
 
 Slide Image 
 1111

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
    
Finally, Julia uses the pagenation function again to get all the slide files, and merges her metadata file with this file information. This way she will be able to review what phenotypes each slide is associated with:

In [26]:
slides = NoadenoJustSlides.researchsubject.file.run()
slidesdf = pd.DataFrame()
for i in slides.paginator(to_df=True):
    slidesdf = pd.concat([slidesdf, i])


Getting results from database

Total execution time: 3522 ms


In [37]:
slidemetadata = pd.merge(slidesdf, 
                         allmetadata, 
                         on="subject_id")
slidemetadata

id  \
0     040229b3-224c-4107-bd33-0854196b6423   
1     040229b3-224c-4107-bd33-0854196b6423   
2     a9e316b2-abcf-4e40-870d-3e1d74abf8e4   
3     a9e316b2-abcf-4e40-870d-3e1d74abf8e4   
4     1595ab35-ad76-4f27-8b2e-8de0482bf164   
...                                    ...   
2145  317ec307-62fd-45d0-b1bd-03aeb094a67e   
2146  0b9e4e31-bbb1-4d88-b50c-a989ce9f4aff   
2147  0b9e4e31-bbb1-4d88-b50c-a989ce9f4aff   
2148  358728ce-157b-4002-8377-5391781a3d57   
2149  358728ce-157b-4002-8377-5391781a3d57   

                                             identifier  \
0     [{'system': 'GDC', 'value': '040229b3-224c-410...   
1     [{'system': 'GDC', 'value': '040229b3-224c-410...   
2     [{'system': 'GDC', 'value': 'a9e316b2-abcf-4e4...   
3     [{'system': 'GDC', 'value': 'a9e316b2-abcf-4e4...   
4     [{'system': 'GDC', 'value': '1595ab35-ad76-4f2...   
...                                                 ...   
2145  [{'system': 'GDC', 'value': '317ec307-62fd-45d...   
2146  [{'system': 'GDC', 'value': '0b9e4e31-bbb1-4d8...   
2147  [{'system': 'GDC', 'value': '0b9e4e31-bbb1-4d8...   
2148  [{'system': 'GDC', 'value': '358728ce-157b-400...   
2149  [{'system': 'GDC', 'value': '358728ce-157b-400...   

                                                  label data_category  \
0     TCGA-VS-A8EI-01Z-00-DX1.8DD9CBFB-C3B2-48D0-ADE...   Biospecimen   
1     TCGA-VS-A8EI-01Z-00-DX1.8DD9CBFB-C3B2-48D0-ADE...   Biospecimen   
2     TCGA-VS-A8EI-01A-01-TS1.64C2A4BF-CE1B-46CB-AE4...   Biospecimen   
3     TCGA-VS-A8EI-01A-01-TS1.64C2A4BF-CE1B-46CB-AE4...   Biospecimen   
4     TCGA-DI-A2QY-11A-01-TS1.9DEA9504-6CF4-4FD2-A91...   Biospecimen   
...                                                 ...           ...   
2145  TCGA-AX-A2H4-01A-02-TSB.E4077E33-365F-46B0-838...   Biospecimen   
2146  TCGA-AX-A2H4-11A-01-TSA.6A13C652-59BA-40D8-937...   Biospecimen   
2147  TCGA-AX-A2H4-11A-01-TSA.6A13C652-59BA-40D8-937...   Biospecimen   
2148  TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...   Biospecimen   
2149  TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...   Biospecimen   

        data_type file_format associated_project  \
0     Slide Image         SVS          TCGA-CESC   
1     Slide Image         SVS          TCGA-CESC   
2     Slide Image         SVS          TCGA-CESC   
3     Slide Image         SVS          TCGA-CESC   
4     Slide Image         SVS          TCGA-UCEC   
...           ...         ...                ...   
2145  Slide Image         SVS          TCGA-UCEC   
2146  Slide Image         SVS          TCGA-UCEC   
2147  Slide Image         SVS          TCGA-UCEC   
2148  Slide Image         SVS          TCGA-CESC   
2149  Slide Image         SVS          TCGA-CESC   

                                                drs_uri  byte_size  \
0     drs://dg.4DFC:040229b3-224c-4107-bd33-0854196b...   54328093   
1     drs://dg.4DFC:040229b3-224c-4107-bd33-0854196b...   54328093   
2     drs://dg.4DFC:a9e316b2-abcf-4e40-870d-3e1d74ab...  192640689   
3     drs://dg.4DFC:a9e316b2-abcf-4e40-870d-3e1d74ab...  192640689   
4     drs://dg.4DFC:1595ab35-ad76-4f27-8b2e-8de0482b...   50173067   
...                                                 ...        ...   
2145  drs://dg.4DFC:317ec307-62fd-45d0-b1bd-03aeb094...  361909653   
2146  drs://dg.4DFC:0b9e4e31-bbb1-4d88-b50c-a989ce9f...  126509613   
2147  drs://dg.4DFC:0b9e4e31-bbb1-4d88-b50c-a989ce9f...  126509613   
2148  drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...  116519647   
2149  drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...  116519647   

                              checksum  ...  \
0     d4ecae7c6f8f467afbcd060058ffca84  ...   
1     d4ecae7c6f8f467afbcd060058ffca84  ...   
2     c9526a0e3df583efda8f0dc61bb21d04  ...   
3     c9526a0e3df583efda8f0dc61bb21d04  ...   
4     79e82fcf3107181102b50ca3e2fde9f5  ...   
...                                ...  ...   
2145  df2c6b4dd80890e4f4bc84dd9c1b44fc  ...   
2146  3b8d3699fad4ea15f5efc8e60e4977e2  ...   
2147  3b8d36

<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
    
This merge seems to have created a lot of duplicate lines. So Julia wants to drop the duplicated lines before saving the dataframe to a csv. Since the identifier columns of the data are lists, python will have a hard time telling if they are duplicated. A quick fix is to subset the data to only the columns that should be used for the deduplication, so Julia first gets the list of all the column names, and then pastes everything but "identifier", "identifier_x", "identifier_y" and "subject_associated_project"into her drop_duplicates call:

In [40]:
slidemetadata.columns

Index(['id', 'identifier', 'label', 'data_category', 'data_type',
       'file_format', 'associated_project', 'drs_uri', 'byte_size', 'checksum',
       'data_modality', 'imaging_modality', 'dbgap_accession_number',
       'imaging_series', 'subject_id', 'researchsubject_id', 'id_x',
       'identifier_x', 'member_of_research_project',
       'primary_diagnosis_condition', 'primary_diagnosis_site', 'id_y',
       'identifier_y', 'species', 'sex', 'race', 'ethnicity', 'days_to_birth',
       'subject_associated_project', 'vital_status', 'days_to_death',
       'cause_of_death'],
      dtype='object')

In [44]:
slidemetadata.drop_duplicates(inplace=True, 
                              subset=['id', 'label', 'data_category', 'data_type',
       'file_format', 'associated_project', 'drs_uri', 'byte_size', 'checksum',
       'data_modality', 'imaging_modality', 'dbgap_accession_number',
       'imaging_series', 'subject_id', 'researchsubject_id', 'id_x',
        'member_of_research_project',
       'primary_diagnosis_condition', 'primary_diagnosis_site', 'id_y',
        'species', 'sex', 'race', 'ethnicity', 'days_to_birth',
         'vital_status', 'days_to_death',
       'cause_of_death'])

In [45]:
slidemetadata

id  \
0     040229b3-224c-4107-bd33-0854196b6423   
2     a9e316b2-abcf-4e40-870d-3e1d74abf8e4   
4     1595ab35-ad76-4f27-8b2e-8de0482bf164   
6     e2086d9b-63ba-4e03-966e-a4427b6ade74   
8     dd8d427d-d63c-41d4-9da7-dab428a3dc3b   
...                                    ...   
2140  9bc76014-1b92-4914-8bf1-482b85bae0dd   
2142  859f8cfb-116d-44ec-9017-b6f6617d62bf   
2144  317ec307-62fd-45d0-b1bd-03aeb094a67e   
2146  0b9e4e31-bbb1-4d88-b50c-a989ce9f4aff   
2148  358728ce-157b-4002-8377-5391781a3d57   

                                             identifier  \
0     [{'system': 'GDC', 'value': '040229b3-224c-410...   
2     [{'system': 'GDC', 'value': 'a9e316b2-abcf-4e4...   
4     [{'system': 'GDC', 'value': '1595ab35-ad76-4f2...   
6     [{'system': 'GDC', 'value': 'e2086d9b-63ba-4e0...   
8     [{'system': 'GDC', 'value': 'dd8d427d-d63c-41d...   
...                                                 ...   
2140  [{'system': 'GDC', 'value': '9bc76014-1b92-491...   
2142  [{'system': 'GDC', 'value': '859f8cfb-116d-44e...   
2144  [{'system': 'GDC', 'value': '317ec307-62fd-45d...   
2146  [{'system': 'GDC', 'value': '0b9e4e31-bbb1-4d8...   
2148  [{'system': 'GDC', 'value': '358728ce-157b-400...   

                                                  label data_category  \
0     TCGA-VS-A8EI-01Z-00-DX1.8DD9CBFB-C3B2-48D0-ADE...   Biospecimen   
2     TCGA-VS-A8EI-01A-01-TS1.64C2A4BF-CE1B-46CB-AE4...   Biospecimen   
4     TCGA-DI-A2QY-11A-01-TS1.9DEA9504-6CF4-4FD2-A91...   Biospecimen   
6     TCGA-DI-A2QY-01Z-00-DX1.EBBED5CC-3098-4694-BCD...   Biospecimen   
8     TCGA-DI-A2QY-01A-01-TS1.62316960-36E9-4919-BE3...   Biospecimen   
...                                                 ...           ...   
2140  TCGA-AJ-A3BG-01A-01-TS1.FDB1B2A0-9AAD-460C-A87...   Biospecimen   
2142  TCGA-AJ-A3BG-01Z-00-DX1.87799C6D-8229-4DE2-BD6...   Biospecimen   
2144  TCGA-AX-A2H4-01A-02-TSB.E4077E33-365F-46B0-838...   Biospecimen   
2146  TCGA-AX-A2H4-11A-01-TSA.6A13C652-59BA-40D8-937...   Biospecimen   
2148  TCGA-EK-A2RB-01A-01-TS1.45101B3C-E301-4BD3-B4E...   Biospecimen   

        data_type file_format associated_project  \
0     Slide Image         SVS          TCGA-CESC   
2     Slide Image         SVS          TCGA-CESC   
4     Slide Image         SVS          TCGA-UCEC   
6     Slide Image         SVS          TCGA-UCEC   
8     Slide Image         SVS          TCGA-UCEC   
...           ...         ...                ...   
2140  Slide Image         SVS          TCGA-UCEC   
2142  Slide Image         SVS          TCGA-UCEC   
2144  Slide Image         SVS          TCGA-UCEC   
2146  Slide Image         SVS          TCGA-UCEC   
2148  Slide Image         SVS          TCGA-CESC   

                                                drs_uri   byte_size  \
0     drs://dg.4DFC:040229b3-224c-4107-bd33-0854196b...    54328093   
2     drs://dg.4DFC:a9e316b2-abcf-4e40-870d-3e1d74ab...   192640689   
4     drs://dg.4DFC:1595ab35-ad76-4f27-8b2e-8de0482b...    50173067   
6     drs://dg.4DFC:e2086d9b-63ba-4e03-966e-a4427b6a...  1283995391   
8     drs://dg.4DFC:dd8d427d-d63c-41d4-9da7-dab428a3...   255850447   
...                                                 ...         ...   
2140  drs://dg.4DFC:9bc76014-1b92-4914-8bf1-482b85ba...    78523868   
2142  drs://dg.4DFC:859f8cfb-116d-44ec-9017-b6f6617d...   185346200   
2144  drs://dg.4DFC:317ec307-62fd-45d0-b1bd-03aeb094...   361909653   
2146  drs://dg.4DFC:0b9e4e31-bbb1-4d88-b50c-a989ce9f...   126509613   
2148  drs://dg.4DFC:358728ce-157b-4002-8377-5391781a...   116519647   

                              checksum  ...  \
0     d4ecae7c6f8f467afbcd060058ffca84  ...   
2     c9526a0e3df583efda8f0dc61bb21d04  ...   
4     79e82fcf3107181102b50ca3e2fde9f5  ...   
6     165a9199090fe0ff9ea86fd2168d41f2  ...   
8     b96eceebd9116cf9567830ce36cd157a  ...   
...                                ...  ...   
2140  6d4195721e8f04d6fbf708fd62f98af3  ...   
2142  67c97665154649a31354b8880aebd713  ...   


<img src="https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true" align="left"
	width="50" height="50" />
    
That looks better! Julia saves this dataframe to a csv, and now she has all the information she needs to begin work on her project. She can use the `drs_id` column information to directly download the images she is interested in using a DRS resolver, or she can input the DRS IDs at a cloud workspace such as [Terra](https://terra.bio/) or the [Cancer Genomics Cloud](https://www.cancergenomicscloud.org/) to view the images online. In either case, she has all the metadata she needs to get started, and can save this notebook of her work in case she'd like to come back and modify her search.

In [28]:
slidemetadata.to_csv("slidemetadata.csv")